File: Java/SingleLinkedList.java
class Node{
	int data;
	Node next;
	Node(int data){
		this.data = data;
		this.next = null;
	}
}
public class SingleLinkedList{
	public Node head;

	void InsertBegin(int data){
		Node temp = new Node(data);
		if (head == null){
			head = temp;
			return;
		}
		temp.next = head;
		head = temp;
	}

	void InsertEnd(int data){
		Node temp = new Node(data);
		if (head == null){
			head = temp;
			return;
		}
		Node trav = head;
		while(trav.next != null){
			trav = trav.next;
		}
		trav.next = temp;
	}
	void InsertBefore(int data,int before){
		Node temp = new Node(data);
		if (head == null || head.data == before){
			temp.next = head;
			head = temp;
			return ;
		}
		Node trav = head.next;
		Node prev = head;
		while(trav != null && trav.data != before){
			prev = trav;
			trav = trav.next;
		}
		prev.next = temp;
		temp.next = trav;
	}
	void InsertAfter(int data,int after){
		Node temp = new Node(data);
		if (head == null ){
			InsertBegin(data);
			return ;
		}
		Node trav = head;
		while(trav.next != null && trav.data != after){
			trav = trav.next;
		}
		temp.next = trav.next;
		trav.next = temp;
	}
	void InsertPosition(int data,int position){
		Node temp = new Node(data);
		if (head == null){
			head = temp;
			return;
		}
		if (position == 1){
			temp.next = head;
			head = temp;
			return;
		}
		Node trav = head;
		int i = 2;
		while(trav.next != null && i < position){
			trav = trav.next;
			i++;
		}
		temp.next = trav.next;
		trav.next = temp;
	}

	void DelBigin(){
		if (head == null){
			return;
		}
		head = head.next;
	}

	void DelEnd(){
		if (head == null){
			return;
		}
		if (head.next == null){
			head = null;
		}
		Node trav = head;
		while(trav.next.next != null){
			trav = trav.next;
		}
		trav.next = null;
	}

	void DelPosition(int position){
		if (head == null){
			return;
		}
		if (position == 1){
			head= head.next ;
			return;
		}
		Node trav = head;
		int i = 2;
		while(trav.next != null && i < position){
			trav = trav.next;
			i++;
		}
		trav.next = trav.next.next;
	}
	void Reverse(){
		if (head == null){
			return;
		}
		Node prev = null;
		Node nextnode;
		Node current = head;
		while(current != null){
			nextnode = current.next;
			current.next = prev;
			prev = current;
			current = nextnode;
		}
		head = prev;
	}
	void Print(){
		Node temp = head;
		while (temp != null) {
			System.out.print(temp.data + " ");
			temp = temp.next;
		}
		System.out.println();
	}

	public static void main(String[] args) {
		SingleLinkedList l = new SingleLinkedList();
		l.InsertBegin(10);
		l.InsertBegin(20);
		l.InsertBegin(30);
		l.InsertBegin(40);
		l.InsertBegin(11);
		l.InsertBegin(23);
		l.InsertBegin(13);
		l.InsertBegin(111);
		l.InsertBegin(333);
		l.Print();
		l.InsertBefore(22,23);
		l.Print();
		l.InsertBefore(332,333);
		l.Print();
		l.InsertAfter(12,11);
		l.Print();
		l.InsertAfter(11,10);
		l.Print();
		// l.DelPosition(9);
		// l.DelBigin();
		// l.DelEnd();
		// l.DelPosition(1);
		// l.DelPosition(4);
		// l.Print();
		// l.Reverse();
		// l.Print();
	}
}

File: CPP/singlelinkedlist.cpp
// implementing single linked list

#include<stdio.h>
#include<stdlib.h>
struct Node{
	int data;
	struct Node* next;
};
struct Node* head;

// Inserting
void InsertAtBegin(int x){
	struct Node* temp = new Node();
	temp->data = x;
	if (head != NULL){
		temp->next = head;
	}
	head = temp;
}
void InsertAtPosition(int x,int p){
	struct Node* temp = new Node();
	temp->data = x;
	if (p==1){
		temp->next = head;
		head = temp;
	}
	else{
		struct Node* trav = head;
		for (int i = 0; i < p-2; ++i)
		{
			trav =  trav->next;
		}
		temp->next = trav->next;
		trav->next = temp;
	}
}
void InsertAtEnd(int x){
	struct Node* temp = new Node();
	temp->data = x;
	if (head == NULL){
		head = temp;
	}
	else{
		struct Node* trav = head;
		while(trav->next != NULL){
			trav = trav->next;
		}
		trav->next = temp;
		temp->next = NULL;
	}
}

// Deleting
void DeleteAtBegin(){
	if (head == NULL){
		printf("Underflow\n");
		return;
	}
	else{
		struct Node* temp = head;
		head = head->next;
		free(temp);
	}
}
void DeleteAtPosition(int p){
	if (head == NULL){
		printf("Underflow\n");
		return;
	}
	struct Node* trav = head;
	if(p ==1){
		head = trav->next;
		free(trav);
		return;
	}
	for (int i = 0; i < p-2; ++i)
	{
		trav = trav->next;
		if (trav->next == NULL){
			printf("Invalid Position\n");
			return;
		}
	}
	struct Node* temp = trav->next; // nth node
	trav->next = temp->next;
	free(temp);
}
void DeleteAtEnd(){
	if (head == NULL){
		printf("Underflow\n");
		return;
	}
	struct Node* trav = head;
	if(head->next == NULL){
		head = NULL;
		free(trav);
	}
	else{
		struct Node* prev = NULL;
		while(trav->next != NULL){
			prev = trav;
			trav = trav->next;
		}
		prev->next = NULL;
		free(trav);
	}
}

// Reverse
void Reverse(){
	struct Node* current = head;
	struct Node* next;
	struct Node* prev = NULL;
	while(current != NULL){
		next = current->next;
		current->next = prev;
		prev = current;
		current = next;
	}
	head = prev;
}
void ReverseRecursive(struct Node* p){
	if (p->next == NULL){
		head = p;
		return;
	}
	ReverseRecursive(p->next);
	struct Node* q = p->next;
	q->next = p;
	p->next = NULL;
}

// Printing
void Print(){
	struct Node* temp = head;
	if (head == NULL){
		printf("List is empty.\n");
		return;
	}
	printf("List is:\n");
	while(temp != NULL){
		printf("%d ", temp->data);
		temp = temp->next;
	}
	printf("\n");
}

void PrintRecursive(struct Node* p){
	if (p == NULL){
		printf("\n");
		return;
	}
	printf("%d ",p->data);
	PrintRecursive(p->next);
}

void PrintReverse(struct Node* p){
	if (p == NULL){
		printf("\n");
		return;
	}
	PrintReverse(p->next);
	printf("%d ",p->data);
}
// Main
int main()
{
	int n,i,x;
	head = NULL;
	InsertAtBegin(2);
	InsertAtEnd(4);
	InsertAtEnd(6);
	InsertAtEnd(5);
	PrintRecursive(head);
	ReverseRecursive(head);
	PrintRecursive(head);
	return 0;
}